File: java/src/main/java/mal/step7_quote.java
package mal;

import java.io.IOException;

import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Iterator;
import mal.types.*;
import mal.readline;
import mal.reader;
import mal.printer;
import mal.env.Env;
import mal.core;

public class step7_quote {
    // read
    public static MalVal READ(String str) throws MalThrowable {
        return reader.read_str(str);
    }

    // eval
    public static Boolean is_pair(MalVal x) {
        return x instanceof MalList && ((MalList)x).size() > 0;
    }

    public static MalVal quasiquote(MalVal ast) {
        if (!is_pair(ast)) {
            return new MalList(new MalSymbol("quote"), ast);
        } else {
            MalVal a0 = ((MalList)ast).nth(0);
            if ((a0 instanceof MalSymbol) &&
                (((MalSymbol)a0).getName() == "unquote")) {
                return ((MalList)ast).nth(1);
            } else if (is_pair(a0)) {
                MalVal a00 = ((MalList)a0).nth(0);
                if ((a00 instanceof MalSymbol) &&
                    (((MalSymbol)a00).getName() == "splice-unquote")) {
                    return new MalList(new MalSymbol("concat"),
                                       ((MalList)a0).nth(1),
                                       quasiquote(((MalList)ast).rest()));
                }
            }
            return new MalList(new MalSymbol("cons"),
                               quasiquote(a0),
                               quasiquote(((MalList)ast).rest()));
        }
    }

    public static MalVal eval_ast(MalVal ast, Env env) throws MalThrowable {
        if (ast instanceof MalSymbol) {
            return env.get((MalSymbol)ast);
        } else if (ast instanceof MalList) {
            MalList old_lst = (MalList)ast;
            MalList new_lst = ast.list_Q() ? new MalList()
                                           : (MalList)new MalVector();
            for (MalVal mv : (List<MalVal>)old_lst.value) {
                new_lst.conj_BANG(EVAL(mv, env));
            }
            return new_lst;
        } else if (ast instanceof MalHashMap) {
            MalHashMap new_hm = new MalHashMap();
            Iterator it = ((MalHashMap)ast).value.entrySet().iterator();
            while (it.hasNext()) {
                Map.Entry entry = (Map.Entry)it.next();
                new_hm.value.put(entry.getKey(), EVAL((MalVal)entry.getValue(), env));
            }
            return new_hm;
        } else {
            return ast;
        }
    }

    public static MalVal EVAL(MalVal orig_ast, Env env) throws MalThrowable {
        MalVal a0, a1,a2, a3, res;
        MalList el;

        while (true) {

        //System.out.println("EVAL: " + printer._pr_str(orig_ast, true));
        if (!orig_ast.list_Q()) {
            return eval_ast(orig_ast, env);
        }

        // apply list
        MalList ast = (MalList)orig_ast;
        if (ast.size() == 0) { return ast; }
        a0 = ast.nth(0);
        String a0sym = a0 instanceof MalSymbol ? ((MalSymbol)a0).getName()
                                               : "__<*fn*>__";
        switch (a0sym) {
        case "def!":
            a1 = ast.nth(1);
            a2 = ast.nth(2);
            res = EVAL(a2, env);
            env.set(((MalSymbol)a1), res);
            return res;
        case "let*":
            a1 = ast.nth(1);
            a2 = ast.nth(2);
            MalSymbol key;
            MalVal val;
            Env let_env = new Env(env);
            for(int i=0; i<((MalList)a1).size(); i+=2) {
                key = (MalSymbol)((MalList)a1).nth(i);
                val = ((MalList)a1).nth(i+1);
                let_env.set(key, EVAL(val, let_env));
            }
            orig_ast = a2;
            env = let_env;
            break;
        case "quote":
            return ast.nth(1);
        case "quasiquote":
            orig_ast = quasiquote(ast.nth(1));
            break;
        case "do":
            eval_ast(ast.slice(1, ast.size()-1), env);
            orig_ast = ast.nth(ast.size()-1);
            break;
        case "if":
            a1 = ast.nth(1);
            MalVal cond = EVAL(a1, env);
            if (cond == types.Nil || cond == types.False) {
                // eval false slot form
                if (ast.size() > 3) {
                    orig_ast = ast.nth(3);
                } else {
                    return types.Nil;
                }
            } else {
                // eval true slot form
                orig_ast = ast.nth(2);
            }
            break;
        case "fn*":
            final MalList a1f = (MalList)ast.nth(1);
            final MalVal a2f = ast.nth(2);
            final Env cur_env = env;
            return new MalFunction (a2f, (mal.env.Env)env, a1f) {
                public MalVal apply(MalList args) throws MalThrowable {
                    return EVAL(a2f, new Env(cur_env, a1f, args));
                }
            };
        default:
            el = (MalList)eval_ast(ast, env);
            MalFunction f = (MalFunction)el.nth(0);
            MalVal fnast = f.getAst();
            if (fnast != null) {
                orig_ast = fnast;
                env = f.genEnv(el.slice(1));
            } else {
                return f.apply(el.rest());
            }
        }

        }
    }

    // print
    public static String PRINT(MalVal exp) {
        return printer._pr_str(exp, true);
    }

    // repl
    public static MalVal RE(Env env, String str) throws MalThrowable {
        return EVAL(READ(str), env);
    }

    public static void main(String[] args) throws MalThrowable {
        String prompt = "user> ";

        final Env repl_env = new Env(null);

        // core.java: defined using Java
        for (String key : core.ns.keySet()) {
            repl_env.set(new MalSymbol(key), core.ns.get(key));
        }
        repl_env.set(new MalSymbol("eval"), new MalFunction() {
            public MalVal apply(MalList args) throws MalThrowable {
                return EVAL(args.nth(0), repl_env);
            }
        });
        MalList _argv = new MalList();
        for (Integer i=1; i < args.length; i++) {
            _argv.conj_BANG(new MalString(args[i]));
        }
        repl_env.set(new MalSymbol("*ARGV*"), _argv);


        // core.mal: defined using the language itself
        RE(repl_env, "(def! not (fn* (a) (if a false true)))");
        RE(repl_env, "(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \")\")))))");
        
        Integer fileIdx = 0;
        if (args.length > 0 && args[0].equals("--raw")) {
            readline.mode = readline.Mode.JAVA;
            fileIdx = 1;
        }
        if (args.length > fileIdx) {
            RE(repl_env, "(load-file \"" + args[fileIdx] + "\")");
            return;
        }
        while (true) {
            String line;
            try {
                line = readline.readline(prompt);
                if (line == null) { continue; }
            } catch (readline.EOFException e) {
                break;
            } catch (IOException e) {
                System.out.println("IOException: " + e.getMessage());
                break;
            }
            try {
                System.out.println(PRINT(RE(repl_env, line)));
            } catch (MalContinue e) {
                continue;
            } catch (MalThrowable t) {
                System.out.println("Error: " + t.getMessage());
                continue;
            } catch (Throwable t) {
                System.out.println("Uncaught " + t + ": " + t.getMessage());
                continue;
            }
        }
    }
}

File: kotlin/src/mal/step7_quote.kt
package mal

fun read(input: String?): MalType = read_str(input)

fun eval(_ast: MalType, _env: Env): MalType {
    var ast = _ast
    var env = _env

    while (true) {
        if (ast is MalList) {
            val first = ast.first()

            if (first is MalSymbol && first.value == "def!") {
                return env.set(ast.nth(1) as MalSymbol, eval(ast.nth(2), env))
            } else if (first is MalSymbol && first.value == "let*") {
                val childEnv = Env(env)
                val bindings = ast.nth(1) as? ISeq ?: throw MalException("expected sequence as the first parameter to let*")

                val it = bindings.seq().iterator()
                while (it.hasNext()) {
                    val key = it.next()
                    if (!it.hasNext()) throw MalException("odd number of binding elements in let*")
                    childEnv.set(key as MalSymbol, eval(it.next(), childEnv))
                }

                env = childEnv
                ast = ast.nth(2)
            } else if (first is MalSymbol && first.value == "fn*") {
                val binds = ast.nth(1) as? ISeq ?: throw MalException("fn* requires a binding list as first parameter")
                val params = binds.seq().filterIsInstance<MalSymbol>()
                val body = ast.nth(2)

                return MalFnFunction(body, params, env, { s: ISeq ->
                    eval(body, Env(env, params, s.seq()))
                })
            } else if (first is MalSymbol && first.value == "do") {
                eval_ast(ast.slice(1, ast.seq().count() - 1), env)
                ast = ast.seq().last()
            } else if (first is MalSymbol && first.value == "if") {
                val check = eval(ast.nth(1), env)

                if (check != NIL && check != FALSE) {
                    ast = ast.nth(2)
                } else if (ast.seq().asSequence().count() > 3) {
                    ast = ast.nth(3)
                } else return NIL
            } else if (first is MalSymbol && first.value == "quote") {
                return ast.nth(1)
            } else if (first is MalSymbol && first.value == "quasiquote") {
                ast = quasiquote(ast.nth(1))
            } else {
                val evaluated = eval_ast(ast, env) as ISeq
                val firstEval = evaluated.first()

                if (firstEval is MalFnFunction) {
                    ast = firstEval.ast
                    env = Env(firstEval.env, firstEval.params, evaluated.rest().seq())
                } else if (firstEval is MalFunction) {
                    return firstEval.apply(evaluated.rest())
                } else throw MalException("cannot execute non-function")
            }
        } else return eval_ast(ast, env)
    }
}

private fun is_pair(ast: MalType): Boolean = ast is ISeq && ast.seq().any()

private fun quasiquote(ast: MalType): MalType {
    if (!is_pair(ast)) {
        val quoted = MalList()
        quoted.conj_BANG(MalSymbol("quote"))
        quoted.conj_BANG(ast)
        return quoted
    }

    val seq = ast as ISeq
    var first = seq.first()

    if ((first as? MalSymbol)?.value == "unquote") {
        return seq.nth(1)
    }

    if (is_pair(first) && ((first as ISeq).first() as? MalSymbol)?.value == "splice-unquote") {
        val spliced = MalList()
        spliced.conj_BANG(MalSymbol("concat"))
        spliced.conj_BANG((first as ISeq).nth(1))
        spliced.conj_BANG(quasiquote(MalList(seq.seq().drop(1).toLinkedList())))
        return spliced
    }

    val consed = MalList()
    consed.conj_BANG(MalSymbol("cons"))
    consed.conj_BANG(quasiquote(ast.first()))
    consed.conj_BANG(quasiquote(MalList(seq.seq().drop(1).toLinkedList())))
    return consed
}

fun eval_ast(ast: MalType, env: Env): MalType =
    if (ast is MalSymbol) {
        env.get(ast)
    } else if (ast is MalList) {
        ast.elements.fold(MalList(), { a, b -> a.conj_BANG(eval(b, env)); a })
    } else if (ast is MalVector) {
        ast.elements.fold(MalVector(), { a, b -> a.conj_BANG(eval(b, env)); a })
    } else if (ast is MalHashMap) {
        ast.elements.entries.fold(MalHashMap(), { a, b -> a.assoc_BANG(b.key, eval(b.value, env)); a })
    } else ast

fun print(result: MalType) = pr_str(result, print_readably = true)

fun rep(input: String, env: Env): String =
        print(eval(read(input), env))

fun main(args: Array<String>) {
    val repl_env = Env()
    ns.forEach({ it -> repl_env.set(it.key, it.value) })

    // Need to cast the strings explicitly to MalType to get this to compile.  Looks like a bug in kotlinc,
    // and it results in a warning.
    repl_env.set(MalSymbol("*ARGV*"), MalList(args.drop(1).map({ it -> MalString(it) as MalType }).toLinkedList()))
    repl_env.set(MalSymbol("eval"), MalFunction({ a: ISeq -> eval(a.first(), repl_env) }))

    rep("(def! not (fn* (a) (if a false true)))", repl_env)
    rep("(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \")\")))))", repl_env)

    if (args.any()) {
        rep("(load-file \"${args[0]}\")", repl_env)
        return
    }

    while (true) {
        val input = readline("user> ")

        try {
            println(rep(input, repl_env))
        } catch (e: EofException) {
            break
        } catch (e: MalContinue) {
        } catch (e: MalException) {
            println("Error: " + e.message)
        } catch (t: Throwable) {
            println("Uncaught " + t + ": " + t.message)
            t.printStackTrace()
        }
    }
}
